### ISSUE TITLE
Unexpected IndexError on systems without NUMA kernel support

### ISSUE THREAD

**Issue by danieldietsch** [sosy-lab/benchexec#440]
When I execute benchexec on an old machine with 
``$ benchexec -N 1 --no-container --maxLogfileSize -1 --no-compress-results  bench.xml``. 
I get the following error:
````
2019-07-12 16:57:38,055 - WARNING - No propertyfile specified. Score computation will ignore the results.
Traceback (most recent call last):
  File "/storage/benchexec/bin/benchexec", line 38, in <module>
    sys.exit(benchexec.benchexec.main())
  File "/storage/benchexec/bin/../benchexec/benchexec.py", line 466, in main
    sys.exit(benchexec.start(argv or sys.argv))
  File "/storage/benchexec/bin/../benchexec/benchexec.py", line 87, in start
    rc = self.execute_benchmark(arg)
  File "/storage/benchexec/bin/../benchexec/benchexec.py", line 351, in execute_benchmark
    result = self.executor.execute_benchmark(benchmark, output_handler)
  File "/storage/benchexec/bin/../benchexec/localexecution.py", line 106, in execute_benchmark
    benchmark.config.coreset,
  File "/storage/benchexec/bin/../benchexec/resources.py", line 98, in get_cpu_cores_per_run
    memory_regions.append(_get_memory_banks_listed_in_dir(coreDir)[0])
IndexError: list index out of range
````

I am using 
  * Python 3.6.5 
  * Gentoo with Kernel 5.1.7 
  * An Intel i7-2600 CPU with 8 (logical) cores and 24GB of memory

I use benchexec from master 11861d9d120e52b312c271fb1438c60c69a9c3ff.

**EDIT** ``python3 -m benchexec.check_cgroups`` executes without output but successful.

**Comment by PhilippWendler** [comment]
Can you please provide the output of `ls -l /sys/devices/system/cpu/cpu*/`? It seems your kernel does not provide the information about the hardware topology that we expect.

Did you maybe compile your kernel with some special settings like disabling NUMA support?

**Comment by danieldietsch** [comment]
Yes
````
/sys/devices/system/cpu/cpu0/:
total 0
drwxr-xr-x 6 root root    0 Jul 14 12:51 cache
lrwxrwxrwx 1 root root    0 Jul 14 05:11 cpufreq -> ../cpufreq/policy0
drwxr-xr-x 7 root root    0 Jul 14 12:51 cpuidle
lrwxrwxrwx 1 root root    0 Jul 14 12:51 driver -> ../../../../bus/cpu/drivers/processor
lrwxrwxrwx 1 root root    0 Jul 14 12:51 firmware_node -> ../../../LNXSYSTM:00/LNXCPU:00
drwxr-xr-x 2 root root    0 Jul 14 12:51 hotplug
drwxr-xr-x 2 root root    0 Jul 14 12:51 microcode
drwxr-xr-x 2 root root    0 Jul 14 12:51 power
lrwxrwxrwx 1 root root    0 Jul 14 12:51 subsystem -> ../../../../bus/cpu
drwxr-xr-x 2 root root    0 Jul 14 12:51 thermal_throttle
drwxr-xr-x 2 root root    0 Jul 14 12:51 topology
-rw-r--r-- 1 root root 4096 Jul 14 12:51 uevent

/sys/devices/system/cpu/cpu1/:
total 0
drwxr-xr-x 6 root root    0 Jul 14 12:51 cache
lrwxrwxrwx 1 root root    0 Jul 14 05:11 cpufreq -> ../cpufreq/policy1
drwxr-xr-x 7 root root    0 Jul 14 12:51 cpuidle
lrwxrwxrwx 1 root root    0 Jul 14 12:51 driver -> ../../../../bus/cpu/drivers/processor
lrwxrwxrwx 1 root root    0 Jul 14 12:51 firmware_node -> ../../../LNXSYSTM:00/LNXCPU:01
drwxr-xr-x 2 root root    0 Jul 14 12:51 hotplug
drwxr-xr-x 2 root root    0 Jul 14 12:51 microcode
-rw-r--r-- 1 root root 4096 Jul 14 12:51 online
drwxr-xr-x 2 root root    0 Jul 14 12:51 power
lrwxrwxrwx 1 root root    0 Jul 14 12:51 subsystem -> ../../../../bus/cpu
drwxr-xr-x 2 root root    0 Jul 14 12:51 thermal_throttle
drwxr-xr-x 2 root root    0 Jul 14 12:51 topology
-rw-r--r-- 1 root root 4096 Jul 14 12:51 uevent

/sys/devices/system/cpu/cpu2/:
total 0
drwxr-xr-x 6 root root    0 Jul 14 12:51 cache
lrwxrwxrwx 1 root root    0 Jul 14 05:11 cpufreq -> ../cpufreq/policy2
drwxr-xr-x 7 root root    0 Jul 14 12:51 cpuidle
lrwxrwxrwx 1 root root    0 Jul 14 12:51 driver -> ../../../../bus/cpu/drivers/processor
lrwxrwxrwx 1 root root    0 Jul 14 12:51 firmware_node -> ../../../LNXSYSTM:00/LNXCPU:02
drwxr-xr-x 2 root root    0 Jul 14 12:51 hotplug
drwxr-xr-x 2 root root    0 Jul 14 12:51 microcode
-rw-r--r-- 1 root root 4096 Jul 14 12:51 online
drwxr-xr-x 2 root root    0 Jul 14 12:51 power
lrwxrwxrwx 1 root root    0 Jul 14 12:51 subsystem -> ../../../../bus/cpu
drwxr-xr-x 2 root root    0 Jul 14 12:51 thermal_throttle
drwxr-xr-x 2 root root    0 Jul 14 12:51 topology
-rw-r--r-- 1 root root 4096 Jul 14 12:51 uevent

/sys/devices/system/cpu/cpu3/:
total 0
drwxr-xr-x 6 root root    0 Jul 14 12:51 cache
lrwxrwxrwx 1 root root    0 Jul 14 05:11 cpufreq -> ../cpufreq/policy3
drwxr-xr-x 7 root root    0 Jul 14 12:51 cpuidle
lrwxrwxrwx 1 root root    0 Jul 14 12:51 driver -> ../../../../bus/cpu/drivers/processor
lrwxrwxrwx 1 root root    0 Jul 14 12:51 firmware_node -> ../../../LNXSYSTM:00/LNXCPU:03
drwxr-xr-x 2 root root    0 Jul 14 12:51 hotplug
drwxr-xr-x 2 root root    0 Jul 14 12:51 microcode
-rw-r--r-- 1 root root 4096 Jul 14 12:51 online
drwxr-xr-x 2 root root    0 Jul 14 12:51 power
lrwxrwxrwx 1 root root    0 Jul 14 12:51 subsystem -> ../../../../bus/cpu
drwxr-xr-x 2 root root    0 Jul 14 12:51 thermal_throttle
drwxr-xr-x 2 root root    0 Jul 14 12:51 topology
-rw-r--r-- 1 root root 4096 Jul 14 12:51 uevent

/sys/devices/system/cpu/cpu4/:
total 0
drwxr-xr-x 6 root root    0 Jul 14 12:51 cache
lrwxrwxrwx 1 root root    0 Jul 14 05:11 cpufreq -> ../cpufreq/policy4
drwxr-xr-x 7 root root    0 Jul 14 12:51 cpuidle
lrwxrwxrwx 1 root root    0 Jul 14 12:51 driver -> ../../../../bus/cpu/drivers/processor
lrwxrwxrwx 1 root root    0 Jul 14 12:51 firmware_node -> ../../../LNXSYSTM:00/LNXCPU:04
drwxr-xr-x 2 root root    0 Jul 14 12:51 hotplug
drwxr-xr-x 2 root root    0 Jul 14 12:51 microcode
-rw-r--r-- 1 root root 4096 Jul 14 12:51 online
drwxr-xr-x 2 root root    0 Jul 14 12:51 power
lrwxrwxrwx 1 root root    0 Jul 14 12:51 subsystem -> ../../../../bus/cpu
drwxr-xr-x 2 root root    0 Jul 14 12:51 thermal_throttle
drwxr-xr-x 2 root root    0 Jul 14 12:51 topology
-rw-r--r-- 1 root root 4096 Jul 14 12:51 uevent

/sys/devices/system/cpu/cpu5/:
total 0
drwxr-xr-x 6 root root    0 Jul 14 12:51 cache
lrwxrwxrwx 1 root root    0 Jul 14 05:11 cpufreq -> ../cpufreq/policy5
drwxr-xr-x 7 root root    0 Jul 14 12:51 cpuidle
lrwxrwxrwx 1 root root    0 Jul 14 12:51 driver -> ../../../../bus/cpu/drivers/processor
lrwxrwxrwx 1 root root    0 Jul 14 12:51 firmware_node -> ../../../LNXSYSTM:00/LNXCPU:05
drwxr-xr-x 2 root root    0 Jul 14 12:51 hotplug
drwxr-xr-x 2 root root    0 Jul 14 12:51 microcode
-rw-r--r-- 1 root root 4096 Jul 14 12:51 online
drwxr-xr-x 2 root root    0 Jul 14 12:51 power
lrwxrwxrwx 1 root root    0 Jul 14 12:51 subsystem -> ../../../../bus/cpu
drwxr-xr-x 2 root root    0 Jul 14 12:51 thermal_throttle
drwxr-xr-x 2 root root    0 Jul 14 12:51 topology
-rw-r--r-- 1 root root 4096 Jul 14 12:51 uevent

/sys/devices/system/cpu/cpu6/:
total 0
drwxr-xr-x 6 root root    0 Jul 14 12:51 cache
lrwxrwxrwx 1 root root    0 Jul 14 05:11 cpufreq -> ../cpufreq/policy6
drwxr-xr-x 7 root root    0 Jul 14 12:51 cpuidle
lrwxrwxrwx 1 root root    0 Jul 14 12:51 driver -> ../../../../bus/cpu/drivers/processor
lrwxrwxrwx 1 root root    0 Jul 14 12:51 firmware_node -> ../../../LNXSYSTM:00/LNXCPU:06
drwxr-xr-x 2 root root    0 Jul 14 12:51 hotplug
drwxr-xr-x 2 root root    0 Jul 14 12:51 microcode
-rw-r--r-- 1 root root 4096 Jul 14 12:51 online
drwxr-xr-x 2 root root    0 Jul 14 12:51 power
lrwxrwxrwx 1 root root    0 Jul 14 12:51 subsystem -> ../../../../bus/cpu
drwxr-xr-x 2 root root    0 Jul 14 12:51 thermal_throttle
drwxr-xr-x 2 root root    0 Jul 14 12:51 topology
-rw-r--r-- 1 root root 4096 Jul 14 12:51 uevent

/sys/devices/system/cpu/cpu7/:
total 0
drwxr-xr-x 6 root root    0 Jul 14 12:51 cache
lrwxrwxrwx 1 root root    0 Jul 14 05:11 cpufreq -> ../cpufreq/policy7
drwxr-xr-x 7 root root    0 Jul 14 12:51 cpuidle
lrwxrwxrwx 1 root root    0 Jul 14 12:51 driver -> ../../../../bus/cpu/drivers/processor
lrwxrwxrwx 1 root root    0 Jul 14 12:51 firmware_node -> ../../../LNXSYSTM:00/LNXCPU:07
drwxr-xr-x 2 root root    0 Jul 14 12:51 hotplug
drwxr-xr-x 2 root root    0 Jul 14 12:51 microcode
-rw-r--r-- 1 root root 4096 Jul 14 12:51 online
drwxr-xr-x 2 root root    0 Jul 14 12:51 power
lrwxrwxrwx 1 root root    0 Jul 14 12:51 subsystem -> ../../../../bus/cpu
drwxr-xr-x 2 root root    0 Jul 14 12:51 thermal_throttle
drwxr-xr-x 2 root root    0 Jul 14 12:51 topology
-rw-r--r-- 1 root root 4096 Jul 14 12:51 uevent

/sys/devices/system/cpu/cpufreq/:
total 0
drwxr-xr-x 2 root root 0 Jul 14 05:11 policy0
drwxr-xr-x 2 root root 0 Jul 14 05:11 policy1
drwxr-xr-x 2 root root 0 Jul 14 05:11 policy2
drwxr-xr-x 2 root root 0 Jul 14 05:11 policy3
drwxr-xr-x 2 root root 0 Jul 14 05:11 policy4
drwxr-xr-x 2 root root 0 Jul 14 05:11 policy5
drwxr-xr-x 2 root root 0 Jul 14 05:11 policy6
drwxr-xr-x 2 root root 0 Jul 14 05:11 policy7

/sys/devices/system/cpu/cpuidle/:
total 0
-r--r--r-- 1 root root 4096 Jul 14 12:51 current_driver
-r--r--r-- 1 root root 4096 Jul 14 12:51 current_governor_ro
````

But you are right, this machine has ``# CONFIG_NUMA is not set`` in the kernel config. But as far as I know, it never had NUMA support and earlier versions of benchexec worked.

**Comment by PhilippWendler** [comment]
Ok, I understand.

The problem was introduced because of a new feature. I think I can fix this, but in general running `benchexec` on systems without `CONFIG_NUMA` is not really supported (and never was, it just worked accidentally) because we cannot test it with reasonable effort. Thus it is possible that further problems occur.

**Comment by danieldietsch** [comment]
Good to know! I can just recompile the kernel of these machines and add NUMA support. But perhaps one could add a note or something to the kernel requirements section in the docs and/or throw a nicer error message.

**Comment by PhilippWendler** [comment]
I did both now. Could you test whether it now works again? (But note that current master already leads to BenchExec 2.0 and has some smaller backwards-incompatible changes).

**Comment by danieldietsch** [comment]
Yes, I tried with d7a9cf1ab

I see the warning, but it seems that together with ``--no-container``, other things break: 
````
../benchexec -N 1 --no-container --maxLogfileSize -1 --no-compress-results  features.xml
2019-07-16 14:58:19,631 - WARNING - No propertyfile specified. Score computation will ignore the results.
2019-07-16 14:58:20,981 - WARNING - Kernel does not have NUMA support. Use benchexec at your own risk.

executing run set 'Automizer'     (251 files)
2019-07-16 14:58:21,017 - INFO - LXCFS is not available, some host information like the uptime leaks into the container.
14:58:21   real-life/BigBranchingExample.bpl                                                                                      2019-07-16 14:58:21,124 - CRITICAL - Cannot execute 'java': Creating namespace for container mode failed: Invalid argument.
FAILED
````

**Comment by PhilippWendler** [comment]
Ouch. I accidentally forced container mode to be always enabled in `benchexec` when I made it the default. It is fixed now, thanks for finding this!

Out of curiosity: Is there a specific reason why you do not use container mode? Even the most loosely configured container (`--network-access --keep-tmp --full-access-dir / --no-tmpfs --keep-system-config`) is still better for isolation than no container, but it is likely that you can do whatever you want successfully with a tighter container. (We could discuss any problems offline or in a separate issue.)

**Comment by danieldietsch** [comment]
Yes, its working for me on aeb48c00b913ed25bedb7469f60b484edeb2ff0f . 

Regarding the container mode: 
  * Its harder to reproduce problems with our tools because we need to ensure that the container itself is not the source of the problem. For example, I cannot simply run the command benchexec used and attach debuggers etc. 
  * I would need to configure the systems I use for container mode. Ultimate usually does not interact with the file system that much, so I do not see the benefit for me. 
  * In those cases where we use the filesystem heavily (i.e., dumping .smt2 scripts), I want to gather them all in one place and not think about having the writes to a ramdisk, as they become huge. 

Come to think of it, half of the time I use benchexec to reliably execute Ultimate in parallel on a machine, and not necessarily to collect the most accurate stats about these executions.
